File: Project-brain/lib/furnace/workCalendar.ts
/**
 * workCalendar.ts — working-time calendar for the browser CPM engine.
 *
 * A deliberate mirror of the backend's app/core/calendar.py. The backend CPM
 * runs in integer *working-day units* and converts units to dates only at the
 * boundary; the browser engine ran the same unit maths but mapped units to
 * dates by plain addition, so every bar drew straight through weekends and
 * holidays. On a five-day calendar a 100-unit activity is 140 calendar days,
 * meaning the client Gantt could show a finish two months earlier than the
 * official backend run for the same schedule.
 *
 * Conventions match the backend exactly (and therefore P6 / MS Project):
 *   - a "unit" is one whole working day
 *   - an activity of duration D starting at unit s occupies s .. s+D-1,
 *     so a one-day task starting Monday also finishes Monday
 *   - working_weekdays uses ISO numbering, Mon=1 .. Sun=7
 *
 * All dates are handled in UTC. Local-time arithmetic would shift a day either
 * side of a DST boundary and quietly move activities by one day twice a year.
 */

export const DAY_MS = 86400000;

export interface WorkCalendarSpec {
  name?: string;
  /** ISO weekday numbers that are working days. Mon=1 .. Sun=7. */
  workingWeekdays?: number[];
  /** ISO date strings ("2026-01-26") that are non-working. */
  holidays?: string[];
  /** ISO date strings forced to be working even if weekend or holiday. */
  exceptionsWork?: string[];
}

const DEFAULT_WEEKDAYS = [1, 2, 3, 4, 5];

/** Midnight UTC of the day containing `ms`. */
function floorDay(ms: number): number {
  return Math.floor(ms / DAY_MS) * DAY_MS;
}

function isoKey(ms: number): string {
  return new Date(floorDay(ms)).toISOString().slice(0, 10);
}

/** ISO weekday (Mon=1 .. Sun=7) for a UTC timestamp. */
function isoWeekday(ms: number): number {
  const jsDay = new Date(ms).getUTCDay();   // Sun=0 .. Sat=6
  return jsDay === 0 ? 7 : jsDay;
}

export function parseDate(value: Date | string | number | null | undefined): number | null {
  if (value === null || value === undefined || value === "") return null;
  if (typeof value === "number") return Number.isFinite(value) ? floorDay(value) : null;
  const ms = value instanceof Date ? value.getTime() : Date.parse(value);
  return Number.isNaN(ms) ? null : floorDay(ms);
}

export class WorkCalendar {
  readonly name: string;
  private readonly weekdays: Set<number>;
  private readonly holidays: Set<string>;
  private readonly exceptions: Set<string>;

  /** unit index -> epoch ms, built lazily and extended on demand. */
  private unitToMs: number[] = [];
  private anchorMs: number | null = null;

  constructor(spec: WorkCalendarSpec = {}) {
    this.name = spec.name ?? "Standard 5-Day";
    const wd = spec.workingWeekdays?.length ? spec.workingWeekdays : DEFAULT_WEEKDAYS;
    this.weekdays = new Set(wd.filter((d) => d >= 1 && d <= 7));
    // A calendar with no working days would hang every stepping loop below.
    if (this.weekdays.size === 0) DEFAULT_WEEKDAYS.forEach((d) => this.weekdays.add(d));
    this.holidays = new Set((spec.holidays ?? []).map((h) => h.slice(0, 10)));
    this.exceptions = new Set((spec.exceptionsWork ?? []).map((h) => h.slice(0, 10)));
  }

  static standard5Day(): WorkCalendar {
    return new WorkCalendar();
  }

  static continuous(): WorkCalendar {
    return new WorkCalendar({ name: "7-Day Continuous", workingWeekdays: [1, 2, 3, 4, 5, 6, 7] });
  }

  isWorkingDay(value: Date | string | number): boolean {
    const ms = parseDate(value);
    if (ms === null) return false;
    const key = isoKey(ms);
    if (this.exceptions.has(key)) return true;
    if (this.holidays.has(key)) return false;
    return this.weekdays.has(isoWeekday(ms));
  }

  firstWorkingOnOrAfter(ms: number): number {
    let d = floorDay(ms);
    for (let guard = 0; guard < 3660 && !this.isWorkingDay(d); guard++) d += DAY_MS;
    return d;
  }

  firstWorkingOnOrBefore(ms: number): number {
    let d = floorDay(ms);
    for (let guard = 0; guard < 3660 && !this.isWorkingDay(d); guard++) d -= DAY_MS;
    return d;
  }

  nextWorkingDay(ms: number): number {
    return this.firstWorkingOnOrAfter(floorDay(ms) + DAY_MS);
  }

  prevWorkingDay(ms: number): number {
    return this.firstWorkingOnOrBefore(floorDay(ms) - DAY_MS);
  }

  /** Set the unit-0 anchor. Snaps forward to the first working day, as the backend does. */
  setAnchor(value: Date | string | number): void {
    const ms = parseDate(value);
    if (ms === null) return;
    const anchor = this.firstWorkingOnOrAfter(ms);
    if (anchor !== this.anchorMs) {
      this.anchorMs = anchor;
      this.unitToMs = [anchor];
    }
  }

  private ensureAnchor(): number {
    if (this.anchorMs === null) this.setAnchor(Date.now());
    return this.anchorMs!;
  }

  /**
   * Date for a working-day unit index. Units may be negative (a constraint or
   * actual date before the anchor), which is why the backward branch exists.
   */
  dateForUnit(unit: number): Date {
    return new Date(this.msForUnit(unit));
  }

  msForUnit(unit: number): number {
    const anchor = this.ensureAnchor();
    const n = Math.round(unit);
    if (n < 0) {
      let ms = anchor;
      for (let i = 0; i < -n; i++) ms = this.prevWorkingDay(ms);
      return ms;
    }
    // cache grows forward only; typical schedules reuse the same prefix
    while (this.unitToMs.length <= n) {
      const last = this.unitToMs[this.unitToMs.length - 1];
      this.unitToMs.push(this.nextWorkingDay(last));
    }
    return this.unitToMs[n];
  }

  /** Working-day unit index for a date, relative to the anchor. */
  unitForDate(value: Date | string | number | null | undefined): number | null {
    const ms = parseDate(value);
    if (ms === null) return null;
    const anchor = this.ensureAnchor();
    if (ms === anchor) return 0;
    if (ms > anchor) {
      // count working days in (anchor, ms]
      let count = 0;
      let cursor = anchor;
      for (let guard = 0; guard < 200000 && cursor < ms; guard++) {
        cursor = this.nextWorkingDay(cursor);
        if (cursor <= ms) count++;
      }
      return count;
    }
    let count = 0;
    let cursor = anchor;
    for (let guard = 0; guard < 200000 && cursor > ms; guard++) {
      cursor = this.prevWorkingDay(cursor);
      if (cursor >= ms) count--;
    }
    return count;
  }

  /**
   * Inclusive count of working days in [start, end]; negative if reversed.
   * Matches the backend's working_days_between.
   */
  workingDaysBetween(
    start: Date | string | number,
    end: Date | string | number,
  ): number {
    const a = parseDate(start);
    const b = parseDate(end);
    if (a === null || b === null) return 0;
    if (a > b) return -this.workingDaysBetween(end, start);
    let count = 0;
    let cursor = a;
    for (let guard = 0; guard < 200000 && cursor <= b; guard++) {
      if (this.isWorkingDay(cursor)) count++;
      cursor += DAY_MS;
    }
    return count;
  }

  /** Date `n` working days from `start`, counting start as unit 0. */
  addWorkingDays(start: Date | string | number, n: number): Date {
    const s = parseDate(start);
    if (s === null) return new Date(NaN);
    let ms = this.firstWorkingOnOrAfter(s);
    const step = n >= 0 ? 1 : -1;
    for (let i = 0; i < Math.abs(Math.round(n)); i++) {
      ms = step > 0 ? this.nextWorkingDay(ms) : this.prevWorkingDay(ms);
    }
    return new Date(ms);
  }

  /**
   * Bar end date for a Gantt, in exclusive-end form.
   *
   * The engine's EF is exclusive (es + duration), but the last worked day is
   * unit ef-1. Drawing a bar to date(ef) would stretch it across whatever
   * non-working days follow, so the bar ends the day after the last worked
   * day and a Friday finish stops at Saturday rather than running to Monday.
   */
  barEndForUnit(efUnit: number): Date {
    const lastWorked = this.msForUnit(Math.max(0, Math.round(efUnit) - 1));
    return new Date(lastWorked + DAY_MS);
  }
}

/** Convenience: build from whatever a schedule payload happens to carry. */
export function calendarFromSchedule(net: {
  workingWeekdays?: number[];
  holidays?: string[];
  exceptionsWork?: string[];
  calendarName?: string;
} | null | undefined): WorkCalendar {
  if (!net) return WorkCalendar.standard5Day();
  return new WorkCalendar({
    name: net.calendarName,
    workingWeekdays: net.workingWeekdays,
    holidays: net.holidays,
    exceptionsWork: net.exceptionsWork,
  });
}

File: project-brain-backend/app/services/work_calendar.py
"""Work calendars for the CPM engine.

A schedule's durations are expressed in *working days*. Without a calendar
the engine treats every calendar day as a working day (the historical
behaviour). A WorkCalendar lets non-working days — weekends and holidays —
be skipped, exactly as Primavera P6 / Synchro do, so computed Early/Late
dates land on real working days and durations mean "N days of work".

Duration convention (matches P6): an activity with duration D that starts on
working day S finishes at the end of the (S + D - 1)-th working day. So a
1-day task starts and finishes the same working day; add_working_days(start, 0)
== start (snapped forward to a working day), and a task's finish is
add_working_days(start, duration - 1).

The default calendar (`SEVEN_DAY`) treats all days as working, which keeps
the engine's behaviour identical to before calendars existed.
"""
from __future__ import annotations

from datetime import date, timedelta
from typing import Iterable, Optional


class WorkCalendar:
    """Working-day calendar.

    working_weekdays: set of Python weekday() ints that are working days
        (Mon=0 .. Sun=6). Default Mon–Sat (P6's common 6-day construction week).
    holidays: set of specific dates that are non-working regardless of weekday.
    """

    def __init__(self, working_weekdays: Optional[Iterable[int]] = None,
                 holidays: Optional[Iterable[date]] = None,
                 name: str = "calendar"):
        self.name = name
        self.working_weekdays = set(
            working_weekdays if working_weekdays is not None else range(6)  # Mon–Sat
        )
        if not self.working_weekdays:
            raise ValueError("A calendar must have at least one working weekday")
        self.holidays = set(holidays or ())

    # ── predicates ────────────────────────────────────────────────
    def is_working(self, d: date) -> bool:
        return d.weekday() in self.working_weekdays and d not in self.holidays

    def next_working(self, d: date) -> date:
        """First working day on or after d."""
        cur = d
        # bounded loop: at most 7 + a holiday run; guard at 400 to be safe
        for _ in range(400):
            if self.is_working(cur):
                return cur
            cur += timedelta(days=1)
        raise ValueError(f"No working day found within 400 days of {d} in calendar {self.name}")

    def prev_working(self, d: date) -> date:
        """Last working day on or before d."""
        cur = d
        for _ in range(400):
            if self.is_working(cur):
                return cur
            cur -= timedelta(days=1)
        raise ValueError(f"No working day found within 400 days before {d} in calendar {self.name}")

    # ── arithmetic ────────────────────────────────────────────────
    def add_working_days(self, d: date, working_days: float) -> date:
        """Advance `working_days` working days from d.

        The start day itself is snapped forward to a working day, then we step
        forward. add_working_days(d, 0) == next_working(d). Negative counts step
        backward (used by finish-driven dependency math)."""
        n = int(round(working_days))
        cur = self.next_working(d) if n >= 0 else self.prev_working(d)
        step = 1 if n >= 0 else -1
        remaining = abs(n)
        while remaining > 0:
            cur += timedelta(days=step)
            if self.is_working(cur):
                remaining -= 1
        return cur

    def finish_from_start(self, start: date, duration_days: float) -> date:
        """Finish date for a task starting `start` with `duration_days` of work.

        1-day task -> finishes same working day. duration D -> start + (D-1)
        working days, on a working day."""
        dur = max(int(round(duration_days)), 1)
        return self.add_working_days(self.next_working(start), dur - 1)

    def start_from_finish(self, finish: date, duration_days: float) -> date:
        """Inverse of finish_from_start: latest start given a finish + duration."""
        dur = max(int(round(duration_days)), 1)
        return self.add_working_days(self.prev_working(finish), -(dur - 1))

    def working_days_between(self, start: date, end: date) -> int:
        """Count working days in [start, end] inclusive of both ends when working.

        Returns a signed count consistent with float math: if end < start the
        result is negative. Used for float = LS - ES in working days."""
        if start == end:
            return 0
        sign = 1 if end >= start else -1
        lo, hi = (start, end) if sign > 0 else (end, start)
        # count working days strictly after lo up to and including hi
        count, cur = 0, lo
        while cur < hi:
            cur += timedelta(days=1)
            if self.is_working(cur):
                count += 1
        return sign * count


# All days work — preserves the engine's pre-calendar behaviour exactly.
SEVEN_DAY = WorkCalendar(working_weekdays=range(7), name="7-day")

# Common Indian construction 6-day week (Sunday off), no holidays by default.
SIX_DAY = WorkCalendar(working_weekdays=range(6), name="6-day")
